utils/package.json

```diff
@@ -25,6 +25,14 @@
     "url": "https://github.com/import-js/eslint-plugin-import/issues"
   },
   "homepage": "https://github.com/import-js/eslint-plugin-import#readme",
+  "peerDependenciesMeta": {
+    "@typescript-eslint/parser": {
+      "optional": true
+    },
+    "eslint-import-resolver-node": {
```

Comment:
This isn’t needed, because it’s already available as a dependency, so the node resolution algorithm means it’s already accessible. There are some broken package managers that require this unnecessary peer dep declaration to make it accessible, but I’m not going to add a breaking change just to cater to broken package managers.

Reply:
`eslint-import-resolver-node` is a dependency of `eslint-plugin-import`, but not of `eslint-module-utils` where it is required by default.

Reply:
eslint-module-utils actually *dynamically* requires `eslint-import-resolver-X`, for every `X` defined in the user's eslint settings.

That the node resolver is the default doesn't mean someone has to use it, and that you can specify a thousand different resolvers (that eslint-module-utils will dutifully try to require) doesn't mean eslint-module-utils should be specifying it.

This kind of dynamic requiring is simply a normal part of the way the entire eslint ecosystem works, and any package manager that breaks it is *itself broken*.

Reply:
Yes, and if a consumer is using a non-default resolver, it is up to them to declare it as a dependency. `eslint-module-utils` is hard coded to use `eslint-import-resolver-node` by default, which is why it makes sense for it to be declared as an _optional_ peer dependency.

Reply:
Correct, this is the right behavior and this causes any projects using yarn v3 to fail without custom configuration.  Perhaps the yarn documentation will make it more clearn as-to why: https://yarnpkg.com/advanced/rulebook#packages-should-only-ever-require-what-they-formally-list-in-their-dependencies

Reply:
@bradennapier that link is an opinion, not a fact, and contradicts the way parts of the npm ecosystem (including eslint itself) have actually worked for over a decade.

Reply:
To be fair, almost all of the ecosystem follows that opinion, and only some plugins of eslint and babel cause trouble...

Reply:
It's a problem for anything that has static configuration, which may include typescript as well as a number of other ecosystems.

Either way, there are valid use cases for not making that convention a constraint.

Reply:
But if `eslint-module-utils` don't declare `eslint-import-resolver-node` as its peer dep, it will cause bugs which you can only fix it by add `eslint-import-resolver-node` to your monorepo project's root path. But that's a very bad practice.

Reply:
@LuckyMona I added the node resolver to my project's devDeps and it seems to just work, no special tricks needed apart from the optional peer dep on eslint.
What error are you seeing?

Reply:
@wmertens hi master,

> "I added the node resolver to my project's devDeps"

Yes, Just only this method can avoid the bug, but which I think is a very bad practice for these reasons:

1. Although this method can work, it's still a phantom dependency case, which we should try to avoid
2. In this case, it can't work: 
    Suppose you developed a npm package, and the package required `eslint-plugin-import` or `eslint-module-utils`, then if  the ancestor consumer of your package uses pnpm, you have to require the ancestor consumer  install `eslint-import-resolver-node` in its devDeps, which almost can't be done. Otherwise the project will report error.

Reply:
Hmm, IMHO the resolver is a plug-in, to be picked by the project. In the situation you describe I would expect that helper package to import the plugin somehow and take care of it, but as a direct consumer I'm responsible for the plugins.

In any case, unless these resolver plugins are the ever-only possible plugins, they should be treated like all plugins. Otherwise you're just moving the problem.

Reply:
Do you mean I should add node resolver to my npm package's devDeps ? But It won't work if you use pnpm, because pnpm's install mechanism forbid phantom dependency, so `eslint-module-utils` can't get node resolver from my npm package and will report error

Reply:
You'd better provide a reproduction instead. AFAIK, pnpm hoists all eslint related dependencies by default.

Reply:
You can try to reproduce it as follows:

One project has A in its Deps, A has B and C in its Deps, C has D in its Deps,  C also require B in its code but don't declare it in its package.json. Then C won't get B from its father A and will report error. (Yarn 3 as the project's package manager and pnpm linker as yarn's nodeLinker)

Given more spare time, I can give a demo.

Reply:
@LuckyMona fwiw this exact kind of conflict is why pnpm/PnP's approach is fundamentally flawed - it's simply not *possible* for this package to explicitly list every plugin package that a user might want to instruct it to load. As such, the only "solution" is either a) hardcoding special cases into their own resolvers, or b) allowing anything you explicitly depend on to be required by anything else you explicitly depend on.

Reply:
Or c) allow specifying plugins by full path.

Reply:
That’s already allowed, but that breaks more than a decade of expectations in the node ecosystem - paths should never be absolute, and node_modules shouldn’t ever have to be hardcoded.

Reply:
Right - that's what require.resolve is for :)

I think we need an example that breaks. I'm in Camp Pure Declarations but plugins cannot be known beforehand so cannot be declared, and AFAIK things just work right now @LuckyMona

Reply:
What is the error you're getting? I don't understand how these settings might be problematic

Reply:
A reproduction is really required here. 😅

Reply:
@LuckyMona eslint config must be able to be static - not JS, but JSON/ini. As such, require.resolve is not available in every case to the config.